src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
+    // method.
+    dwResult = CallEHFilterFunclet(OBJECTREFToObject(throwable),
+                                  GetFrameRestoreBase(pvRegDisplay->pCallerContext),
+                                  CastHandlerFn(pfnHandler),
+                                  pFuncletCallerSP);
+#else
+    dwResult = pfnHandler(establisherFrame, OBJECTREFToObject(throwable));
+#endif        
+
+    // Profiler, debugger and ETW events
+    MakeCallbacksRelatedToHandler(false, pThread, pMD, &pExInfo->m_CurrentClause, (DWORD_PTR)pFilterIP, GetSP(pvRegDisplay->pCurrentContext));
+    END_QCALL;
+
+    return dwResult == EXCEPTION_EXECUTE_HANDLER;
+}
+
+
+struct ExtendedEHClauseEnumerator : EH_CLAUSE_ENUMERATOR
+{
+    StackFrameIterator *pFrameIter;
+    unsigned EHCount;
+};
+
+extern "C" BOOL QCALLTYPE RhpEHEnumInitFromStackFrameIterator(StackFrameIterator *pFrameIter, BYTE** pMethodStartAddress, EH_CLAUSE_ENUMERATOR * pEHEnum)
+{
+    QCALL_CONTRACT;
+
+    ExtendedEHClauseEnumerator *pExtendedEHEnum = (ExtendedEHClauseEnumerator*)pEHEnum;
+    pExtendedEHEnum->pFrameIter = pFrameIter;
+
+    BEGIN_QCALL;
+    Thread* pThread = GET_THREAD();
+    Frame* pFrame = pThread->GetFrame();
+    MarkInlinedCallFrameAsEHHelperCall(pFrame);
+
+    IJitManager* pJitMan = pFrameIter->m_crawl.GetJitManager();
+    const METHODTOKEN& MethToken = pFrameIter->m_crawl.GetMethodToken();
+    *pMethodStartAddress = (BYTE*)pJitMan->JitTokenToStartAddress(MethToken);
+    pExtendedEHEnum->EHCount = pJitMan->InitializeEHEnumeration(MethToken, pEHEnum);
+
+    END_QCALL;
+
+    return pExtendedEHEnum->EHCount != 0;
+}
+
+extern "C" BOOL QCALLTYPE RhpEHEnumNext(EH_CLAUSE_ENUMERATOR* pEHEnum, RhEHClause* pEHClause)
+{
+    QCALL_CONTRACT;
+    BOOL result = FALSE;
+
+    BEGIN_QCALL;
+    Thread* pThread = GET_THREAD();
+    Frame* pFrame = pThread->GetFrame();
+    MarkInlinedCallFrameAsEHHelperCall(pFrame);
+
+    ExtendedEHClauseEnumerator *pExtendedEHEnum = (ExtendedEHClauseEnumerator*)pEHEnum;
+    StackFrameIterator *pFrameIter = pExtendedEHEnum->pFrameIter;
+
+    if (pEHEnum->iCurrentPos < pExtendedEHEnum->EHCount)
+    {
+        IJitManager* pJitMan   = pFrameIter->m_crawl.GetJitManager();
+        const METHODTOKEN& MethToken = pFrameIter->m_crawl.GetMethodToken();
+
+        EE_ILEXCEPTION_CLAUSE EHClause;
+        memset(&EHClause, 0, sizeof(EE_ILEXCEPTION_CLAUSE));
+        PTR_EXCEPTION_CLAUSE_TOKEN pEHClauseToken = pJitMan->GetNextEHClause(pEHEnum, &EHClause);
+        Thread* pThread = GET_THREAD();
+        ExInfo* pExInfo = pThread->GetExceptionState()->GetCurrentExInfo();
+        pExInfo->m_CurrentClause = EHClause;
+
+        pEHClause->_tryStartOffset = EHClause.TryStartPC;
+        pEHClause->_tryEndOffset = EHClause.TryEndPC;
+        if (IsFilterHandler(&EHClause))
+        {
+            pEHClause->_filterAddress =  (BYTE*)pJitMan->GetCodeAddressForRelOffset(MethToken, EHClause.FilterOffset);
```

Comment:
`GetCodeAddressForRelOffset` is an expensive call when the code is hot/cold split. Can we delay it to only once we need the address like it is done in the current exception handling?

Reply:
That would mean diverging between NativeAOT and the new EH semantics of the `_filterAddress` field though.

Reply:
We can update NativeAOT to do the same. The NativeAOT semantics are not necessarily the best unified semantics.

This can be `TODO-NewEH`.

Reply:
Added TODO-NewEH